src/main/scala/firrtl/Parser.scala

```diff
@@ -42,12 +44,26 @@ object Parser extends LazyLogging {
       parser.getInterpreter.setPredictionMode(PredictionMode.SLL)
       parser.addParseListener(listener)
 
-      // Concrete Syntax Tree
-      parser.circuit
+      // Syntax errors may violate assumptions in the Listener and Visitor
+      // We need to handle these errors gracefully
+      val exception =
+        try {
+          parser.circuit
+          None
+        } catch {
+          case e: UnsupportedVersionException => throw e
+          case NonFatal(e) => Some(e)
+        }
 
       val numSyntaxErrors = parser.getNumberOfSyntaxErrors
       if (numSyntaxErrors > 0) throw new SyntaxErrorsException(s"$numSyntaxErrors syntax error(s) detected")
 
+      // Any exceptions thrown by the Listener/Visitor should be due to syntax errors reported
+      // above, but we don't want to mask any bugs.
+      if (exception.nonEmpty) {
+        throw exception.get
+      }
```

Comment:
Nit: `if (foo) { throw bar }` is inconsistent with the style on line 59.

Reply:
Is this code path ever expected to be hit or is this a complete fallback that "something went wrong"? This seems like the latter. If that is the case, it may be useful to add more information to this error that it looks like a real FIRRTL bug. Maybe this is already reported as such.

Reply:
> Nit: if (foo) { throw bar } is inconsistent with the style on line 59.

Fixed

> Is this code path ever expected to be hit or is this a complete fallback that "something went wrong"? This seems like the latter. If that is the case, it may be useful to add more information to this error that it looks like a real FIRRTL bug. Maybe this is already reported as such.

It is the latter. I've made the comment call that out more clearly and wrapped the issue in an "Internal Exception". Such wrapping was already occurting if you hit it from `FirrtlPhase`, but since `FirrtlPhase` is smart enough not to wrap twice, might as well wrap it here so that if you call the firrtl Parser directly, you still get the wrapping.